### DISCUSSION TITLE
Yolov8 Queue Management get queue counts

### DISCUSSION THREAD

**Discussion by cobaltautomationdev** [ultralytics/ultralytics#10260] [Q&A]
how can I get the value of queue counts with Queue Management ? it always get 0. based on your sourcing code.

```
 if len(self.reg_pts) >= 3:
                is_inside = self.counting_region.contains(Point(track_line[-1]))
                if prev_position is not None and is_inside:
                    self.counts += 1

    label = "Queue Counts : " + str(self.counts)
```

```
self.counts = 0
self.display_frames()
```

**Comment by glenn-jocher** [comment]
Hello! Thanks for reaching out with your question on queue counts in YOLOv8 Queue Management. Based on what you've shared, it appears there might be an issue with how points are being detected or classified within your specified queue region. The condition `if len(self.reg_pts) >= 3` suggests that the region is being set, but the crux likely relates to how `is_inside` evaluates `Point(track_line[-1])`.

To help guide you in the right direction, consider the following common areas to check:
- Ensure `self.reg_pts` correctly defines the counting region polygon. Verify the points make up a coherent shape that represents your counting area accurately.
- Make sure `Point(track_line[-1])` represents the object's position in the frame correctly. The point might not be updating as expected, or it might not lie within your defined region.
- Confirm there are objects moving through the counting region and that their movement aligns with the logic set for incrementing `self.counts`.

If your setup seems correct on these fronts, you might want to add print statements or logging to verify `is_inside` is `True` when expected. This can pinpoint if the issue lies in the detection or in the logic checking the region.

Remember, the counting logic increments the queue count only if an object's latest point is inside the counting region and there was a previous position defined.

Still have issues? Feel free to share more details or any specific behavior you observe. Happy coding! 😊

**Reply by cobaltautomationdev** [reply]
Hi glenn， Thanks for you kindly help. But you misunderstand my question. I just want to know how to get the  queue counts after run your Queue Management.  The code is what I copied from your sourcing code in your document.
![image](https://github.com/ultralytics/ultralytics/assets/145321686/37536f0d-d008-4f9e-be3b-d7aa1c7f2ff1)

**Reply by glenn-jocher** [reply]
Hello! I understand you're looking to obtain queue counts after running the Queue Management feature with YOLOv8. Thanks for clarifying your query. 😊 

From the code snippet you provided and considering the typical flow of our Queue Management system, the queue counts should update as objects (people, vehicles, etc.) pass through a defined region. If you're following our guidelines and have integrated the snippet correctly, obtaining queue counts should be straightforward. These counts are stored in `self.counts`.

Here's a quick guide to ensure you're checking the counts correctly:

1. Ensure your counting regions (`self.reg_pts`) are correctly defined and encompass the area where queueing occurs.
2. Confirm that `self.counts` is being updated in your code as objects move through the counting region—this variable holds the queue count.

After running the system, if you need to access the queue count, simply refer to `self.counts`. For example:

```python
print(f"Current queue count: {self.counts}")
```

If the counts remain at 0, verify that objects are indeed passing through the defined region and that their last tracked point falls within this area.

For further assistance, don't hesitate to reach out. Happy coding! 😊

**Reply by cobaltautomationdev** [reply]
The instance variable self.counts within your extract_and_process_tracks(tracks) method ultimately gets reset to 0. My objective is to mimic the behavior of an object counter, specifically managing separate counts for incoming (in_counts) and outgoing (out_counts).

In pursuit of this, I have made adjustments to your source code by relocating the assignment self.counts = 0 to a more appropriate position. Now, to retrieve the current count statistics associated with the registration points, one simply needs to invoke queue.counts. This modification facilitates straightforward access to the updated counts directly from the queue instance.

 ```
 if tracks[0].boxes.id is not None:
            boxes = tracks[0].boxes.xyxy.cpu()
            clss = tracks[0].boxes.cls.cpu().tolist()
            track_ids = tracks[0].boxes.id.int().cpu().tolist()
            self.counts = 0
```

**Reply by glenn-jocher** [reply]
Hi there! 👋 Your approach to adjusting the source code for managing incoming (`in_counts`) and outgoing (`out_counts`) queues sounds like a solid strategy. Modifying `self.counts` initialization placement to better suit your counting logic is a thoughtful way to customize the behavior of the object counter. To further enhance your solution and make counts for incoming and outgoing objects separately trackable, you might consider creating two separate instance variables, say `self.in_counts` and `self.out_counts`. Then, based on the direction or condition you define for distinguishing incoming from outgoing, you could increment these counters accordingly. Here's a simplified example:

```python
if condition_for_incoming:
    self.in_counts += 1
elif condition_for_outgoing:
    self.out_counts += 1
```

This way, `self.in_counts` and `self.out_counts` give you direct access to the separate counts, which you can retrieve at any point. Happy coding! 😊
